architecture-decision-records/0001-multi-account-terraform.md

```diff
@@ -0,0 +1,55 @@
+# 1. Use Terraform across multiple AWS accounts
+
+**Date:** 2020-02-20
+
+The actual decision was made earlier than this, between December 2019 and
+January 2020.
+
+## Context
+
+We have created multiple AWS accounts to host the different TDR environments,
+and we're happy with the decision we made during the Alpha to use Terraform to
+configure the infrastructure (see [terraform.md][alpha-terraform] and
+[terraform_vs_cloudformation.md][tf-vs-cf]).
+
+The Alpha only ran in one AWS account. We used Terraform workspaces so that we
+could create multiple environments if we wanted, but in the end we only used one
+environment for prototyping and user testing.
+
+In Alpha, we ran Terraform scripts from our dev machines. For Beta, we'd prefer
+to run Terraform in a controlled environment like Jenkins. This has a few
+advantages:
+
+- Makes it easier to control access to production infrastructure: a developer
+  can have permission to run Jenkins jobs (and therefore make
+  version-controlled, peer-reviewed changes to production) without having
+  credentials giving them full access to the production account
+- Makes it easier to deploy changes to each environment in turn
+- Provides a record of (recent) Terraform runs, and who started them
+
+But there's a chicken-and-egg problem with running Terraform from a hosted
+system like Jenkins, because it would be helpful if we could use Terraform to
+create the Jenkins infrastructure itself.
+
+[alpha-terraform]: ../technology-considerations/terraform.md
+[tf-vs-cf]: ../technology-considerations/terraform_vs_cloudformation.md
+
+## Decision
+
+Create multiple Terraform projects, which are run in different stages:
+
+- A backend project which is run against the management account and bootstraps
+  the rest of the infrastructure. It is run from a development machine and
+  creates the IAM roles and Terraform state storage (S3 and DynamoDB) that will
+  be used by the environment-specific Terraform scripts. This project has no
+  shared Terraform state file.
```

Comment:
The backend project does have a shared state file now because we needed different developers to be able to work on it.  The S3 bucket and Dynamo table for it were created manually.

Reply:
Fixed, thanks.